### PR TITLE
Fix edge handling and arrow type

### LINKED ISSUE #1687: Arrow of class diagram inheritance arrow missing
There are two options. 
* Introduce another arrow like --i> , --[>, --o>  or  --|:> 
* Change the black arrow --|> to be white

What do you think

### LINKED ISSUE #3817: Bug: type issue in ts@4.8+ with esm
### Description

The package is shipping esm files, but package.json is missing "type": "module"

So typescript will assume users are importing a commonjs module, so it will report issues with:

```ts
import memaid from "mitt";

memaid.someAPI(...args)
```

Saying the expression is not callable.

That's because a cjs module will have it's module.exports on default key, so:

```ts
import memaid from "mitt";

memaid.default.someAPI(...args)
```

is ok to typescript, but it's not acually working (because we do have esm modules using export default) and the default key does not exisit.

### Steps to reproduce

Create project with `"type": "module"` and install mermaid. then ensure ts.config.json has `target: "ESxxx"`, you will get issues mention above. (Does not need to use `moduleResolution: NodeNext`)

### Screenshots

_No response_

### Code Sample

_No response_

### Setup

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jgreywolf**
## :bookmark_tabs: Summary

The main intent of this work is to ensure that we are using a "hollow" arrowhead for inheritance and implementation as detailed in UML specs - https://en.wikipedia.org/wiki/Class_diagram#/media/File:Uml_classes_en.svg

Resolves #3817
Resolves #1687

## :straight_ruler: Design Decisions

The biggest issue here is being able to draw a arrowhead with no fill.  Doing such creates an arrowhead with a line in the middle. In this case the idea is to determine what the background color is, and fill the arrowhead with the same color - or white if the bg is transparent or not detected.

This requires walking up the DOM to get to an element outside of the SVG itself that has the background color property set.  Depending on where the diagram is in any page this could involve a lot of recursion

As such, I am slightly concerned about potential performance hit

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [ ] :computer: have added unit/e2e tests (if appropriate)
- [ ] :notebook: have added documentation (if appropriate)
- [ ] :bookmark: targeted `develop` branch

### CODE REVIEW

File: packages/mermaid/src/dagre-wrapper/markers.js [hunk 1]
```diff
@@ -11,6 +48,8 @@ const insertMarkers = (elem, markerArray, type, id) => {
 
 const extension = (elem, type, id) => {
   log.trace('Making markers for ', id);
+  let backgroundColor = getBackgroundColor(elem);
```
Comment (aloisklink) [inline packages/mermaid/src/dagre-wrapper/markers.js:51]:
```suggestion
  const backgroundColor = getBackgroundColor(elem);
```

This should be `const` if we never change the `backgroundColor` variable.

**Review by aloisklink — COMMENTED** [review]
A couple of issues (although they seem pretty minor, so I wouldn't mind if it's just documented somewhere that these are purposely not covered to make implementation easier):
  - What if the background isn't using the CSS `background-color` property, but is instead using something like `background: green` or `background-image: url(...)`?
  - What happens if the background is a transparent color (e.g. `rgba(255, 0, 0, 0.2)`?

Ideally, we'd figure out how to stop drawing the line under the arrow head, but I'm not an SVG expert, so I've got no idea how hard that would be. It might not be worth the effort just to handle a couple of rare edge cases :shrug:

**Comment by sidharthv96** [comment]
Tangent: Should we add some performance benchmarks?
Something that integrates into imgsnapshot tests would make sure most cases are covered, but we'll also need some tests with multiple iterations.

File: packages/mermaid/src/dagre-wrapper/markers.js [hunk 2]
```diff
@@ -2,6 +2,43 @@
...
+  if (container._groups) {
+    container = container._groups[0][0];
+  }
+  if (container.tagName.toLowerCase() === 'g') {
+    container = container.parentElement;
+  }
+  if (container.localName.toLowerCase() === 'svg') {
+    container = container.parentElement;
+  }
+  return container;
```
Comment (sidharthv96) [inline packages/mermaid/src/dagre-wrapper/markers.js:20]:
```suggestion
  // the intent here is to find the first parent element that is NOT part of the SVG element
  // I know there has to be a better way, but could not find one that worked
  // tried using checking if elem was instanceof SVGGraphicsElement or SVGElement, but it failed to detect correctly
  if (container._groups) {
    return container._groups[0][0];
  }
  if (container.tagName.toLowerCase() === 'g') {
    return container.parentElement;
  }
  if (container.localName.toLowerCase() === 'svg') {
    return container.parentElement;
  }
  return elem;
```

**Review by sidharthv96 — COMMENTED** [review]
Like Alois said, stopping the line before the arrowhead is the best solution if possible.

File: packages/mermaid/src/dagre-wrapper/markers.js (hunk 1 above)
Comment (jgreywolf) [reply packages/mermaid/src/dagre-wrapper/markers.js:51]: you are correct - originally it would change, but I did not adjust when implementation was improved

File: packages/mermaid/src/dagre-wrapper/markers.js (hunk 2 above)
Comment (jgreywolf) [reply packages/mermaid/src/dagre-wrapper/markers.js:20]: We cant make this change. If container._groups then we need to get the first element in the first element - but that wont get us to the correct element. That just gets us to an element that has a "parentElement" property. So, it still needs the other checks as it goes.  If we are in a "g" element we want the parent, and after THAT, we want the svg element.

I know there is a better way to do this...

**Comment by jgreywolf** [comment]
> A couple of issues (although they seem pretty minor, so I wouldn't mind if it's just documented somewhere that these are purposely not covered to make implementation easier):
> 
>     * What if the background isn't using the CSS `background-color` property, but is instead using something like `background: green` or `background-image: url(...)`?
> 
>     * What happens if the background is a transparent color (e.g. `rgba(255, 0, 0, 0.2)`?
> 
> 
> Ideally, we'd figure out how to stop drawing the line under the arrow head, but I'm not an SVG expert, so I've got no idea how hard that would be. It might not be worth the effort just to handle a couple of rare edge cases 🤷

It doesnt matter if the background isnt using a CSS property.  the code is grabbing the "computed" styles - so if there is a background at all this method will pull it out.
And if it is transparent I am just setting it to white


As to not drawing the line.  Totally agreed. I started looking into this and have not figure out how to adjust this yet

**Comment by aloisklink** [comment]
> And if it is transparent I am just setting it to white

I think the code currently only checks for 100% white transparency, aka `rgba(0, 0, 0, 0)`, see https://github.com/mermaid-js/mermaid/blob/65c0c0acb727880e0416bd3e3ba8d119ec70a82a/packages/mermaid/src/dagre-wrapper/markers.js#L39

I'm not sure what it does if you have partial transparencies (e.g. `rgba(255, 0, 0, 0.2)`, which is a red with 80% transparency (aka pink if the background is white)).

> As to not drawing the line. Totally agreed. I started looking into this and have not figure out how to adjust this yet

If it's too difficult, this background hack should still handle 99.9% of cases. We just need to document somewhere that this is a hack, why the hack exists, and the limitations it has. E.g. something like:

```js
/**
 * Finds the parent background color of an SVG element.
 *
 * Used to make a "hollow" arrowhead for an arrow.
 *
 * We can't use a transparent fill,
 * because the arrow line is behind the arrowhead
 * (ideally we'd stop drawing the line behind the arrowhead,
 *  but this is pretty complicated to do). 
 *
 * **Limitations**:
 * - If the parent background color is a partial transparency,
 *   the arrowhead will also be partially transparent.
 * - More complicated backgrounds like pictures/animations won't work.
 */
const getBackgroundColor = (elem) => {
```

We could always figure out not drawing the line under the arrowheads in the future.

Honestly, most class UML diagrams just pick white for hollow arrows, so picking a background color is definitely and improvement, e.g. like:

image

File: packages/mermaid/src/dagre-wrapper/markers.js (hunk 1 above)
Comment (aloisklink) [reply packages/mermaid/src/dagre-wrapper/markers.js]: > you are correct - originally it would change

We should probably add some sort of ESlint rule that automatically catches this and fixes this for us.

File: packages/mermaid/src/dagre-wrapper/markers.js (hunk 2 above)
Comment (sidharthv96) [reply packages/mermaid/src/dagre-wrapper/markers.js:20]: Ahh, my bad. Those were `if`s, not `else if`s...

**Comment by jgreywolf** [comment]
Didnt realize this was still pending

File: packages/mermaid/src/dagre-wrapper/markers.js [hunk 3]
```diff
@@ -2,6 +2,59 @@
...
+ * We can't use a transparent fill,
+ * because the arrow line is behind the arrowhead
+ * (ideally we'd stop drawing the line behind the arrowhead,
+ *  but this is pretty complicated to do).
+ *
+ * **Limitations**:
+ * - If the parent background color is a partial transparency,
+ *   the arrowhead will also be partially transparent.
+ * - More complicated backgrounds like pictures/animations won't work.
+ * @param elem
```
Comment (aloisklink) [inline packages/mermaid/src/dagre-wrapper/markers.js:37]: Can you add a short description for the `elem` parameter (it's mainly to make `eslint`/`tsdoc` lint checkers happy)?

Maybe something like:

```suggestion
 * @param elem The SVG element.
```

**Review by aloisklink — APPROVED** [review]
Looks good to me!

Is it worth adding some of limitations described in https://github.com/mermaid-js/mermaid/pull/4582#issuecomment-1622266331 into the PR description?

File: packages/mermaid/src/dagre-wrapper/markers.js [hunk 4]
```diff
@@ -22,7 +77,8 @@ const extension = (elem, type, id) => {
     .attr('markerHeight', 240)
     .attr('orient', 'auto')
     .append('path')
-    .attr('d', 'M 1,7 L18,13 V 1 Z');
+    .attr('d', 'M 1,7 L18,13 V 2 Z')
```
Comment (sidharthv96) [inline packages/mermaid/src/dagre-wrapper/markers.js:80]: I'm not familiar with SVG, so why is `V 1 Z` changed to `V 2 Z` ?

**Review by sidharthv96 — APPROVED** [review]

**Comment by sidharthv96** [comment]
image
image


In dark mode, white looks wrong. What can be done here?

**Comment by sidharthv96** [comment]
By tweaking the linefunction and refX, we can actually achive proper transparent markers. 
So, I think we can avoid the DOM walking, and the white fallback. 

image

**Comment by jgreywolf** [comment]
> By tweaking the linefunction and refX, we can actually achive proper transparent markers. It will require some extra handling of cases, but I think we can avoid the DOM walking, and the white fallback.

But what determines those are the right values to use?  magic numbers...

**Comment by sidharthv96** [comment]
The values to use needs to be computed for the marker types. It's based on the Ref values they currently have. 
The only remaining problem is when nodes have an angle. Let me see if that can be handled.

**Comment by sidharthv96** [comment]
image

Almost done, a little tweak needed for minor adjustment of the offset.

File: packages/mermaid/src/diagrams/class/classDiagram.spec.ts [hunk 5]
```diff
@@ -1193,16 +1193,18 @@ describe('given a class diagram with relationships, ', function () {
     });
 
     it('should handle relation definitions with type only on right side', function () {
-      const str = 'classDiagram\n' + 'Class1 --|> Class02';
+      const str = 'classDiagram\n' + 'Class1 --|> Class2';
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/class/classDiagram.spec.ts:1196]: We already support Extension `--|>` and Realization `..|>`. 
Wouldn't releasing this change mark the extension as a realization, changing the meaning of a lot of existing diagrams?

```mermaid
classDiagram
classA <|-- classB
classM <|.. classN
```

**Review by sidharthv96 — CHANGES_REQUESTED** [review]

**Comment by jgreywolf** [comment]
I am going to close this PR, and redo to take into account the fix with https://github.com/mermaid-js/mermaid/pull/4788
